Sources/NIOCore/IPAddresses.swift

```diff
@@ -0,0 +1,577 @@
+            }
+        }
+        
+        /// Creates a new `IPv4Address`.
+        ///
+        /// - parameters:
+        ///   - address: Bytes that hold the IPv4 address.
+        public init(address: IPv4Bytes) {
+            self.address = address
+        }
+        
+        /// Creates a new `IPv4Address`.
+        ///
+        /// - parameters:
+        ///   - packedBytes: Collection of UInt8 that holds the address.
+        public init<Bytes: Collection>(packedBytes bytes: Bytes) where Bytes.Element == UInt8 {
+            var ipv4Bytes = IPv4Bytes((0,0,0,0))
+            
+            for (idx, elt) in bytes.prefix(4).enumerated() {
+                ipv4Bytes[idx] = elt
+            }
+            
+            self = .init(address: ipv4Bytes)
+        }
+        
+        /// Creates a new `IPv4Address`.
+        ///
+        /// - parameters:
+        ///   - string: String representation of an IPv4 address.
+        public init(string: String) throws {
+            var bytes: [UInt8] = [0,0,0,0]
+            var idx: Int = 0
+            
+            for char in string.utf8 {
+                if char == UInt8(ascii: ".") {
+                    idx += 1
+                } else if UInt8(ascii: "0")...UInt8(ascii: "9") ~= char {
+                    let number = char - UInt8(ascii: "0")
+                    if idx > 3 || bytes[idx] > 25 || (255 - number < (bytes[idx] * 10)) {
+                        throw IPAddressError.failedToParseIPString(string)
+                    }
+                    bytes[idx] = bytes[idx] * 10 + number
+                } else {
+                    throw IPAddressError.failedToParseIPString(string)
+                }
+            }
+            if idx != 3 {
+                throw IPAddressError.failedToParseIPString(string)
+            }
+            self = .init(packedBytes: bytes)
+        }
+    }
+    
+    /// A single IPv6 address for `IPAddress`
+    public struct IPv6Address: CustomStringConvertible {
+        /// The bytes storing the address of the IPv6 address.
+        public var address: IPv6Bytes
+        
+        /// Get the `IPv6Address` as a string.
+        public var ipAddressString: String {
+            let bucketValues = stride(from: 0, to: 15, by: 2).lazy.map({ idx in
+                UInt16(uint8Tuple: (self.address[idx], self.address[idx + 1]))
+            })
+            
+            var bucketStrings: Array<String> = bucketValues.map {String(uint16Hex: $0)}
+            
+            // find most consecutive zeros and remove them
+            var consecutiveZeros = [0, 0, 0, 0, 0, 0, 0, 0]
+            var consecutiveCounter = 0
+            for (idx, value) in bucketValues.enumerated() {
+                switch value {
+                case 0: consecutiveCounter += 1
+                default: consecutiveCounter = 0
+                }
+                consecutiveZeros[idx] = consecutiveCounter
+            }
```

Comment:
Do we need the Array? Presumably we need to keep track of the last best candidate, and then our current state, and if at any point our current state ends up better than the last best candidate we can overwrite it, right?